### DISCUSSION TITLE
Publish locales as separate packages

### DISCUSSION THREAD

**Discussion by pkuczynski** [faker-js/faker#1199] [Polls]
Example usage proposal:

```js
import { faker } from '@faker/faker-js' // faker logic + en
import { de } from '@faker/faker-js-de' // only definitions

faker.locale = de
faker.firstName()

faker.locale = null
faker.firstName()    // en
```

Pros:
* smaller package size, faster download, and installation, less carbon footprint
* user can choose what he needs exactly
* side effect: we will get some statistics on which locales are most popular

Cons:
* user needs to specify every locale he needs manually (package.json, imports, faker instantiation)

**Comment by ST-DDT** [comment]
That's what tree shaking is for?!

**Reply by pkuczynski** [reply]
How could tree shaking work if we allow to do `faker.locale = 'de'`?

**Reply by ST-DDT** [reply]
Sorry to be nitpicky but thats not what the question has asked. I know what this question is supposed to be about, but its not explicitly mentioned.
If we use the `locale/pl` variant, then we use tree shaking because all other locales get dropped.
We still "cannot" do `faker.locale = 'de'` there.

Note even if you have separate packages, this does not affect whether `faker.locale = 'de'` is possible or not.

**Reply by pkuczynski** [reply]
As this is a spin-off from https://github.com/faker-js/faker/discussions/1049#discussioncomment-3255454, I understand it was not very clear when viewed without it. So I updated the desc to include example usage. Is it clearer now @ST-DDT?

**Reply by ST-DDT** [reply]
Maybe I should rephrase my question as well.

What benefits do separate packages offer over simple tree shaking?

**Reply by pkuczynski** [reply]
Tree shaking would only come into play when you include faker in your build. Then (hopefully) all unused code will be removed from the build of your package. But then, I guess, people rather do not include faker in their builds, as faker is a tool used in CI, and not in the production code?

Right now we publish only single package with all locales included. Npm site says it has an unpacked size of 8.94 MB. 80-90% of it are locales. When I run local build for example, I can see that `dist/cjs` folder weights 11M, where `dist/cjs/modules` is 152K and `dist/cjs/ocales` are 10M. So if we would publish `@faker/faker-js` with pure modules + en, the size would drop to few hundred K, and each locale would be published separately. That means users and CI would not need to download stuff they never use.

Imagine you would install only `@faker/faker-js` (modules + en) and `@faker/faker-js-de`, in total something like 200K instead of downloading and installing 9MB, of which you only need small part.

Does that make sense now?

Note: _package sizes are only examples to express the idea_

**Comment by xDivisionByZerox** [comment]
> ... less carbon footprint

LOL, I never see this mentioned in the tech community as a valid argument. Thanks for giving it the weight and validity it deserves.
